File: src/main/scala/com/exasol/cloudetl/parquet/ParquetValueConverter.scala
package com.exasol.cloudetl.parquet

import java.util.List

import com.exasol.common.json.JsonMapper
import com.exasol.parquetio.data.Row

import org.apache.parquet.schema.LogicalTypeAnnotation.uuidType
import org.apache.parquet.schema.MessageType
import org.apache.parquet.schema.PrimitiveType
import org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY
import org.apache.parquet.schema.Type.Repetition
import java.util.UUID

/**
 * A Parquet value converter class that transforms nested values to JSON strings.
 */
final case class ParquetValueConverter(schema: MessageType) {
  private[this] val size = schema.getFields.size()
  private[this] var convertedValues = Array.ofDim[Object](size)

  /**
   * Converts {@link Row} into array of values.
   *
   * It maps the complex types (e.g, {@code MAP}, {@code LIST}) into JSON strings. The result array is emitted as an
   * Exasol table row.
   *
   * @param row a Parquet row
   * @return converted array of values
   */
  def convert(row: Row): Array[Object] = convertParquetComplexValuesToJSON(row.getValues())

  private[this] def convertParquetComplexValuesToJSON(values: List[Object]): Array[Object] = {
    var i = 0
    while (i < size) {
      convertedValues(i) = convertValue(i, values.get(i))
      i += 1
    }
    convertedValues
  }

  private[this] def convertValue(i: Int, value: Object): Object = {
    val fieldType = schema.getType(i)
    if (fieldType.isPrimitive() && !fieldType.isRepetition(Repetition.REPEATED)) {
      convertPrimitiveValue(fieldType.asPrimitiveType(), value)
    } else {
      JsonMapper.toJson(value)
    }
  }

  private[this] def convertPrimitiveValue(primitiveType: PrimitiveType, value: Object): Object = {
    val logicalType = primitiveType.getLogicalTypeAnnotation()
    primitiveType.getPrimitiveTypeName() match {
      case FIXED_LEN_BYTE_ARRAY if logicalType == uuidType() => value.asInstanceOf[UUID].toString()
      case _                                                 => value
    }
  }

}

File: src/main/java/com/exasol/cloudetl/parquet/ParquetValueConverter.java
package com.exasol.cloudetl.parquet;

import java.util.List;
import java.util.Objects;
import java.util.UUID;

import org.apache.parquet.schema.LogicalTypeAnnotation;
import org.apache.parquet.schema.MessageType;
import org.apache.parquet.schema.PrimitiveType;
import org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName;
import org.apache.parquet.schema.Type.Repetition;

import com.exasol.common.json.JsonMapper;
import com.exasol.parquetio.data.Row;

/** Converts Parquet rows to Exasol-emittable values. */
public final class ParquetValueConverter {
    private final MessageType schema;
    private final int size;
    private final Object[] convertedValues;

    /** Create a converter. */
    public ParquetValueConverter(final MessageType schema) {
        this.schema = schema;
        this.size = schema.getFields().size();
        this.convertedValues = new Object[this.size];
    }

    /** Factory for Java callers. */
    public static ParquetValueConverter apply(final MessageType schema) {
        return new ParquetValueConverter(schema);
    }

    /** Convert a row. */
    public Object[] convert(final Row row) {
        return convertParquetComplexValuesToJson(row.getValues());
    }

    private Object[] convertParquetComplexValuesToJson(final List<Object> values) {
        for (int i = 0; i < this.size; i++) {
            this.convertedValues[i] = convertValue(i, values.get(i));
        }
        return this.convertedValues;
    }

    private Object convertValue(final int index, final Object value) {
        final org.apache.parquet.schema.Type fieldType = this.schema.getType(index);
        if (fieldType.isPrimitive() && !fieldType.isRepetition(Repetition.REPEATED)) {
            return convertPrimitiveValue(fieldType.asPrimitiveType(), value);
        }
        return JsonMapper.toJson(value);
    }

    private Object convertPrimitiveValue(final PrimitiveType primitiveType, final Object value) {
        final LogicalTypeAnnotation logicalType = primitiveType.getLogicalTypeAnnotation();
        if (primitiveType.getPrimitiveTypeName() == PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY
                && Objects.equals(logicalType, LogicalTypeAnnotation.uuidType())) {
            return ((UUID) value).toString();
        }
        return value;
    }
}
